### PR TITLE
Shifting of graph from Dashboard to Service layer

### PR DESCRIPTION
**Pull request by Rajat-Dabade**
Close https://github.com/SigNoz/signoz/issues/3024

### CODE REVIEW

File: frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+				{!isEmpty(widget) && prevChartDataSetRef?.labels ? (
+					<>
+						<div className="drag-handle">
+							<WidgetHeader
+								parentHover={hovered}
+								title={widget?.title}
+								widget={widget}
+								onView={handleOnView}
+								onDelete={(): void => {}}
+								onClone={(): void => {}}
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx:241]: we are breaking dry principle here

File: frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+	onDragSelect?: (start: number, end: number) => void;
+	onClickHandler?: GraphOnClickHandler;
+}
+
+GraphWithoutDashboard.defaultProps = {
+	onDragSelect: undefined,
+	onClickHandler: undefined,
+};
+
+const mapDispatchToProps = (
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx:345]: if there is no dashboard why we are importing actions against dashboard

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx [hunk 3]
```diff
@@ -119,25 +125,37 @@ function WidgetHeader({
...
-				label: 'Clone',
-			},
-			{
-				key: keyMethodMapping.delete.key,
-				icon: <DeleteOutlined />,
-				disabled: !deleteWidget,
-				danger: true,
-				label: 'Delete',
-			},
+			...(allowEdit
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:128]: this is not scalable

Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:128]: this is not scalable

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx [hunk 4]
```diff
@@ -119,25 +125,37 @@ function WidgetHeader({
...
+				? [
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:129]:
```
// Define all possible actions.
const actions = [
  { key: 'view', icon: FullscreenOutlined, label: 'View', condition: true, disabled: () => queryResponse.isLoading },
  { key: 'edit', icon: EditFilled, label: 'Edit', condition: () => allowEdit, disabled: () => !editWidget },
  { key: 'clone', icon: CopyOutlined, label: 'Clone', condition: () => allowClone, disabled: () => !editWidget },
  { key: 'delete', icon: DeleteOutlined, label: 'Delete', condition: () => allowDelete, disabled: () => !deleteWidget, danger: true },
];

const menuList = useMemo(() => {
  return actions
    .filter(action => action.condition)
    .map(({ key, icon: Icon, label, ...rest }) => {
      return {
        key: keyMethodMapping[key].key,
        icon: <Icon />,
        label,
        ...rest
      };
    });
}, [deleteWidget, editWidget, keyMethodMapping, queryResponse.isLoading, allowClone, allowDelete, allowEdit]);
```

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx (hunk 2 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx:345]: Ah, I forgot to remove it. Will remove it.

File: frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx (hunk 1 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx:241]: Okay, I will make it optional in WidgetHeader.

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx (hunk 3 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx]: Got it

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx (hunk 4 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx]: Okay, will make the changes.

**Review by palashgdev — CHANGES_REQUESTED** [review]

**Comment by Rajat-Dabade** [comment]
This PR should immediately after https://github.com/SigNoz/signoz/pull/3039

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx [hunk 5]
```diff
@@ -111,45 +125,64 @@ function WidgetHeader({
...
-	const menuList: MenuItemType[] = useMemo(
-		() => [
+	const actions = useMemo(
+		(): MenuItem[] => [
 			{
-				key: keyMethodMapping.view.key,
+				key: MenuItemKeys.View,
 				icon: <FullscreenOutlined />,
+				label: MenuItemLabels.View,
+				condition: true,
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:134]: what is the meaning of condition

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx [hunk 6]
```diff
@@ -111,45 +125,64 @@ function WidgetHeader({
...
 			editWidget,
-			keyMethodMapping.delete.key,
-			keyMethodMapping.edit.key,
-			keyMethodMapping.view.key,
-			keyMethodMapping.clone.key,
-			queryResponse.isLoading,
 		],
 	);
 
+	const menuList: MenuItemType[] = useMemo(
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:170]: make a util for this one

File: frontend/src/container/MetricsApplication/MetricsPageQueries/DBCallQueries.ts [hunk 7]
```diff
@@ -12,35 +12,42 @@ export const databaseCallsRPS = ({
...
 }: DatabaseCallsRPSProps): QueryBuilderData => {
-	const metricName: BaseAutocompleteData = {
-		dataType: 'float64',
-		isColumn: true,
-		key: 'signoz_db_latency_count',
-		type: null,
-	};
+	const metricNames: BaseAutocompleteData[] = [
+		{
+			dataType: 'float64',
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/DBCallQueries.ts:17]: a lot of magic strings are here, please make constant for same

File: frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts [hunk 8]
```diff
@@ -5,39 +5,63 @@ import {
...
+	aggregateOperator,
+	dataSource,
+	queryName,
+	expression,
 }: BuilderQueriesProps): QueryBuilderData => ({
 	queryFormulas: [],
-	queryData: [
-		{
+	queryData: metricNames.map((item, index) => {
+		const obj = {
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts:30]: what is the meaning of obj ?

File: frontend/src/container/MetricsApplication/Tabs/External.tsx [hunk 9]
```diff
@@ -106,6 +112,11 @@ function External(): JSX.Element {
 		[servicename, tagFilterItems],
 	);
 
+	externalCallErrorWidget.title = EXTERNAL_CALL_ERROR_PERCENTAGE;
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/External.tsx:115]: can we directly use the title here ?

File: frontend/src/container/MetricsApplication/Tabs/constant.ts [hunk 10]
```diff
@@ -1,3 +1,14 @@
 export const legend = {
 	address: '{{address}}',
 };
+
+export const LATENCY = 'Latency';
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/constant.ts:5]: is this title ?

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx (hunk 5 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:134]: Changed the name

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx (hunk 6 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:170]: Okay

File: frontend/src/container/MetricsApplication/Tabs/constant.ts (hunk 10 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/constant.ts]: Will change this.

File: frontend/src/container/MetricsApplication/Tabs/External.tsx (hunk 9 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/External.tsx:115]: Okay, I figured it out, where we can use this.

File: frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts (hunk 8 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts:30]: Will change it

File: frontend/src/container/MetricsApplication/MetricsPageQueries/DBCallQueries.ts (hunk 7 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/DBCallQueries.ts]: Sure

File: frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+import { AppState } from 'store/reducers';
+import { Widgets } from 'types/api/dashboard/getAll';
+import { GlobalReducer } from 'types/reducer/globalTime';
+
+import EmptyWidget from '../EmptyWidget';
+import WidgetHeader from '../WidgetHeader';
+import FullView from './FullView/index.metricsBuilder';
+import { FullViewContainer, Modal } from './styles';
+
+function GraphWithoutDashboard({
```
[the remark this answers is not in the document]
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/Graph/GraphWithoutDashboard.tsx]: Removed the component

File: frontend/src/container/GridGraphLayout/WidgetHeader/contants.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+export enum MenuItemKeys {
+	View = 'view',
+	Edit = 'edit',
+	Delete = 'delete',
+	Clone = 'clone',
+}
+
+export enum MenuItemLabels {
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/contants.ts:8]: mapping between key and label should be there

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx [hunk 13]
```diff
@@ -200,4 +214,12 @@ function WidgetHeader({
 	);
 }
 
+WidgetHeader.defaultProps = {
+	onDelete: undefined,
+	onClone: undefined,
+	allowDelete: undefined,
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:220]: two default value in above this is true and here it is undefined

File: frontend/src/container/GridGraphLayout/WidgetHeader/types.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+	isVisible: boolean;
+	disabled: boolean;
+	danger?: boolean;
+}
+
+export type TWidgetOptions =
+	| MenuItemKeys.View
+	| MenuItemKeys.Edit
+	| MenuItemKeys.Delete
+	| string;
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/types.ts:18]: string should be there

File: frontend/src/container/GridGraphLayout/WidgetHeader/utils.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import { MenuItemType } from 'antd/es/menu/hooks/useItems';
+
+import { MenuItem, TWidgetOptions } from './types';
+
+export const generateMenuList = (
+	actions: MenuItem[],
+	keyMethodMapping: {
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/WidgetHeader/utils.ts:7]: duplicate types

File: frontend/src/container/GridGraphLayout/Graph/index.tsx [hunk 16]
```diff
@@ -77,9 +85,15 @@ function GridCardGraph({
 	const { dashboards } = useSelector<AppState, DashboardReducer>(
 		(state) => state.dashboards,
 	);
-	const [selectedDashboard] = dashboards;
-	const selectedData = selectedDashboard?.data;
-	const { variables } = selectedData;
+
+	let selectedDashboard: Dashboard | undefined;
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/index.tsx:89]: make a util for this one which give selectedDashboard for this one

File: frontend/src/container/MetricsApplication/Tabs/Overview/ServiceOverview.tsx [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+					}),
+					clickhouse_sql: [],
+					id: uuid(),
+				},
+				GraphTitle.LATENCY,
+			),
+		[servicename, tagFilterItems],
+	);
+
+	const { isLoading } = useQuery<PayloadProps>({
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/Tabs/Overview/ServiceOverview.tsx:62]: what's the sense for this calling this useQuery if we are not consuming this data

File: frontend/src/container/MetricsApplication/constant.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+	ERROR_PERCENTAGE = 'Error Percentage',
+	DATABASE_CALLS_RPS = 'Database Calls RPS',
+	DATABASE_CALLS_AVG_DURATION = 'Database Calls Avg Duration',
+	EXTERNAL_CALL_ERROR_PERCENTAGE = 'External Call Error Percentage',
+	EXTERNAL_CALL_DURATION = 'External Call duration',
+	EXTERNAL_CALL_RPS_BY_ADDRESS = 'External Call RPS(by Address)',
+	EXTERNAL_CALL_DURATION_BY_ADDRESS = 'External Call duration(by Address)',
+}
+
+export enum OPERATOR {
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/constant.ts:26]: duplicate types

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/GridGraphLayout/WidgetHeader/contants.ts (hunk 12 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/contants.ts:8]: Okay

File: frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx (hunk 13 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/WidgetHeader/index.tsx:220]: Will change it to true.

File: frontend/src/container/GridGraphLayout/Graph/index.tsx (hunk 16 above)
Comment (Rajat-Dabade) [reply frontend/src/container/GridGraphLayout/Graph/index.tsx:89]: Okay

File: frontend/src/container/MetricsApplication/Tabs/Overview/ServiceOverview.tsx (hunk 17 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/Tabs/Overview/ServiceOverview.tsx:62]: Yes, agreed. Will remove this.

File: frontend/src/container/GridGraphLayout/Graph/index.tsx [hunk 19]
```diff
@@ -43,6 +48,7 @@ import WidgetHeader from '../WidgetHeader';
 import FullView from './FullView/index.metricsBuilder';
 import { FullViewContainer, Modal } from './styles';
 
+// eslint-disable-next-line sonarjs/cognitive-complexity
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/index.tsx:51]: let remove this one

File: frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts [hunk 20]
```diff
@@ -5,39 +5,59 @@ import {
...
 			reduceTo: 'sum',
 			filters: {
-				items: itemsA,
+				items: filterItems[index],
 				op: 'AND',
 			},
-		},
-	],
+			dataSource,
+		} as IBuilderQuery;
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts:52]: don't enforce the types

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 21]
```diff
@@ -1,55 +1,150 @@
...
+	OPERATION_LEGENDS,
+	QUERYNAME_AND_EXPRESSION,
+	WidgetKeys,
+} from '../constant';
 import {
 	getQueryBuilderQueries,
 	getQueryBuilderQuerieswithFormula,
 } from './MetricsPageQueriesFactory';
 
+export const letency = ({
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:21]: `latency`

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 22]
```diff
@@ -1,55 +1,150 @@
...
+	servicename,
+	tagFilterItems,
+}: LetencyProps): QueryBuilderData => {
+	const metricNames: BaseAutocompleteData[] = [
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:25]: no relation between names and types

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts [hunk 23]
```diff
@@ -1,55 +1,150 @@
...
+	QUERYNAME_AND_EXPRESSION,
+	WidgetKeys,
+} from '../constant';
 import {
 	getQueryBuilderQueries,
 	getQueryBuilderQuerieswithFormula,
 } from './MetricsPageQueriesFactory';
 
+export const letency = ({
+	servicename,
```
Comment (palashgdev) [inline frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:22]: serviceName should be a string here as backend would throw error when service name is undefined

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts (hunk 20 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/MetricsPageQueriesFactory.ts:52]: Sure, will remove this

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 21 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:21]: Okay

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 22 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:25]: Will change the name

File: frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts (hunk 23 above)
Comment (Rajat-Dabade) [reply frontend/src/container/MetricsApplication/MetricsPageQueries/OverviewQueries.ts:22]: `serviceName` is coming from Params and can be undefined. So I think there should be a check of it.

**Comment by palashgdev** [comment]
css for header is breaking

File: frontend/src/container/GridGraphLayout/Graph/index.tsx [hunk 24]
```diff
@@ -374,10 +395,18 @@ interface GridCardGraphProps extends DispatchProps {
...
+	onClickHandler?: GraphOnClickHandler;
+	allowDelete?: boolean;
+	allowClone?: boolean;
+	allowEdit?: boolean;
 }
 
 GridCardGraph.defaultProps = {
 	onDragSelect: undefined,
+	onClickHandler: undefined,
+	allowDelete: undefined,
```
Comment (palashgdev) [inline frontend/src/container/GridGraphLayout/Graph/index.tsx:407]: default values should be true

**Review by palashgdev — CHANGES_REQUESTED** [review]

**Review by palashgdev — APPROVED** [review]

**Comment by makeavish** [comment]
@Rajat-Dabade Please fix frontend build pipeline failure

**Comment by Rajat-Dabade** [comment]
@makeavish I have updated yAxisUnit and now it showing in nanoseconds

Screenshot 2023-07-17 at 10 56 17 PM

**Review by makeavish — APPROVED** [review]

**Review by palashgdev — APPROVED** [review]
Great @Rajat-Dabade 👏 and thanks for addressing the changes quick :-D

**Pull request merged**